src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Packaging/ManagedProjectSystemPackage.cs

```diff
@@ -20,73 +16,32 @@ namespace Microsoft.VisualStudio.Packaging
     [PackageRegistration(AllowsBackgroundLoading = true, RegisterUsing = RegistrationMethod.CodeBase, UseManagedResourcesOnly = true)]
     [ProvideProjectFactory(typeof(XprojProjectFactory), null, "#27", "xproj", "xproj", null)]
     [ProvideAutoLoad(ActivationContextGuid, PackageAutoLoadFlags.BackgroundLoad)]
-    [ProvideUIContextRule(ActivationContextGuid, "Load Managed Project Package",
-        "dotnetcore",
-        new string[] { "dotnetcore" },
-        new string[] { "SolutionHasProjectCapability:.NET & CPS" }
+    [ProvideUIContextRule(ActivationContextGuid,
+        name: "Load Managed Project Package",
+        expression: "dotnetcore",
+        termNames: new[] { "dotnetcore" },
+        termValues: new[] { "SolutionHasProjectCapability:.NET & CPS" }
         )]
-
     [ProvideMenuResource("Menus.ctmenu", 4)]
-    internal partial class ManagedProjectSystemPackage : AsyncPackage
+    internal sealed class ManagedProjectSystemPackage : AsyncPackage
     {
         public const string ActivationContextGuid = "E7DF1626-44DD-4E8C-A8A0-92EAB6DDC16E";
         public const string PackageGuid = "860A27C0-B665-47F3-BC12-637E16A1050A";
 
-        private IDotNetCoreProjectCompatibilityDetector? _dotNetCoreCompatibilityDetector;
-        private IVsRegisterProjectSelector? _projectSelectorService;
-        private uint _projectSelectorCookie = VSConstants.VSCOOKIE_NIL;
-
-        public ManagedProjectSystemPackage()
-        {
-        }
-
         protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
 #pragma warning disable RS0030 // Do not used banned APIs
-            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-
-            _projectSelectorService = await this.GetServiceAsync<SVsRegisterProjectTypes, IVsRegisterProjectSelector>();
+            Assumes.True(ThreadHelper.JoinableTaskContext.IsOnMainThread, "Must be on UI thread.");
```

Comment:
This is an incorrect assertion, see: http://index/?query=AsyncPackage.InitializeAsync&rightProject=Microsoft.VisualStudio.Shell.15.0&file=AsyncPackage.cs&line=69, in particular:

```
This method is invoked from a background thread.
```

Reply:
See `LoadType::Asynchronous` in vspakmgr.cpp.

Reply:
I didn't see this on a background thread, but indeed the code suggests it might.

> a UI context rule

Could you elaborate on that?

Reply:
Packages can be initialized for a variety of reasons, due to a synchronous load due a project type or something else referring to it, an explicit load by other code, or due to UI contexts. UI contexts are trigger points that force a package to be loaded, such as "Solution has projects", "Solution has C# projects", "Debugging", etc. We have a custom UI context which you can see applied to the package that says if there are any projects with CPS or .NET capabilities.

Reply:
Thanks, that's great to know.

Debugging today I noticed that the package init was happening much earlier than it was last week.